hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/health/TestEmptyContainerHandler.java

```diff
@@ -0,0 +1,194 @@
+        ecReplicationConfig, 1, CLOSING, 0L, 123L);
+
+    // though key count is 0, the container and its replicas are not CLOSED
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            ContainerReplicaProto.State.OPEN, 0L, 123L, 1, 2, 3, 4, 5);
+
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.emptyList())
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    assertAndVerify(request, false, 0);
+  }
+
+  /**
+   * This test exists to verify that the definition of an empty container is
+   * 0 key count. Number of used bytes are not considered.
+   */
+  @Test
+  public void testNonEmptyRatisContainerReturnsFalse() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ratisReplicationConfig, 1, CLOSED, 5L, 123L);
+
+    // though container and its replicas are CLOSED, key count is not 0
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            ContainerReplicaProto.State.CLOSED, 5L, 123L, 0, 0, 0);
+
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.emptyList())
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    assertAndVerify(request, false, 0);
+  }
+
+  @Test
+  public void testEmptyContainerWithNonZeroBytesUsedReturnsTrue() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ratisReplicationConfig, 1, CLOSED, 0L, 123L);
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            ContainerReplicaProto.State.CLOSED, 0L, 123L, 0, 0, 0);
+
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.emptyList())
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    assertAndVerify(request, true, 3);
+  }
+
+  private void assertAndVerify(ContainerCheckRequest request,
+      boolean assertion, int times) {
+    Assertions.assertEquals(assertion, emptyContainerHandler.handle(request));
+    try {
+      Mockito.verify(replicationManager, Mockito.times(times))
+          .sendDeleteCommand(Mockito.any(ContainerInfo.class), Mockito.anyInt(),
+              Mockito.any(DatanodeDetails.class));
+
+      if (times > 0) {
+        Mockito.verify(containerManager, Mockito.times(1))
+            .updateContainerState(Mockito.any(ContainerID.class),
+                Mockito.any(HddsProtos.LifeCycleEvent.class));
+      }
+    } catch (Exception e) {
```

Comment:
Why do we need to catch an exception here?

Reply:
Alright, I'll let it propagate up.